File: app/src/main/java/com/szeptun/shoppinglist/ui/ShoppingList/ShoppingListViewModel.java
package com.szeptun.shoppinglist.ui.ShoppingList;

import com.szeptun.shoppinglist.domain.SaveShoppingList;
import com.szeptun.shoppinglist.entity.ShoppingList;

import javax.inject.Inject;

public class ShoppingListViewModel {
    @Inject
    SaveShoppingList saveShoppingList;

    @Inject
    public ShoppingListViewModel() {
    }

    public void saveList(ShoppingList shoppingList) {
        saveShoppingList.execute(shoppingList);
    }
}

File: app/src/main/java/com/szeptun/shoppinglist/ui/ShoppingList/ShoppingListViewModel.kt
package com.szeptun.shoppinglist.ui.ShoppingList

import androidx.lifecycle.ViewModel
import com.szeptun.shoppinglist.domain.RemoveProduct
import com.szeptun.shoppinglist.domain.SaveShoppingList
import com.szeptun.shoppinglist.entity.Product
import com.szeptun.shoppinglist.entity.ShoppingList
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.rxkotlin.addTo
import javax.inject.Inject

class ShoppingListViewModel @Inject constructor(
    private val saveShoppingList: SaveShoppingList,
    private val removeProduct: RemoveProduct
) : ViewModel() {

    private val onClearDisposable = CompositeDisposable()

    fun saveList(shoppingList: ShoppingList) {
        saveShoppingList.execute(shoppingList)
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe {}
            .addTo(onClearDisposable)
    }

    fun removeProduct(product: Product) {
        removeProduct.execute(product)
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe()
            .addTo(onClearDisposable)
    }

    override fun onCleared() {
        super.onCleared()
        onClearDisposable.clear()
    }
}
